modules/assim.batch/R/pda.generate.externals.R

```diff
@@ -0,0 +1,189 @@
+##' @param external.knots boolean, if TRUE pass prior.list, ind.list, nknots OR knots.list arguments too
+##' @param knots.list a list of dataframes (one per pft) where each row is a parameter vector, i.e. training points for the emulator. 
+##' If not NULL these are used, otherwise knots will be generated using prior.list, ind.list and nknots.
+##' @param ind.list a named list of vectors (one per pft), where each vector indicates the indices of the parameters on the prior.list targeted in the PDA 
+##' e.g. ind.list <-  list(temperate.deciduous = c(2), temperate.conifer = c(1,2))
+##' @param nknots number of knots you want to train the emulator on
+##' @export
+##' example
+##'
+##' pda.externals <-  pda.generate.externals(external.data   = TRUE, obs = obs, varn = "NEE", varid = 297, n_eff = 106.9386, external.formats = TRUE,
+##'                                         model_data_diag = TRUE, model.out = "/data/workflows/PEcAn_15000000111/out/15000186876",
+##'                                         start_date = "2017-01-01", end_date = "2018-12-31")
+
+pda.generate.externals <-  function(external.data    = FALSE, obs = NULL, varn = NULL, varid = NULL, n_eff = NULL, align_method = "match_timestep", par = NULL,
+                                    model_data_diag  = FALSE, model.out = NULL, start_date = NULL, end_date = NULL,
+                                    external.formats = FALSE,
+                                    external.priors  = FALSE, prior.list = NULL,
+                                    external.knots   = FALSE, knots.list = NULL, ind.list = NULL, nknots = NULL){
+  
+  pda.externals <- list()
+  ##################### external.data #####################
+  if(external.data){
+    if(is.null(obs) & is.null(varn) & is.null(varid)){
+      stop("If you want to generate external.data, the following args cannot be NULL: obs, varn, varid")
+    }
+    external.data <-  vector("list", length(varn)) 
+    for(i in seq_along(external.data)){
+      
+      # fill in external.data sublists : variable.name
+      variable.name <-  list(variable.name = list(variable.drv = varn[i], variable.eqn = list(variables = varn[i], expression = varn[i])))
+      external.data[[i]]$variable.name <-  variable.name
+      
+      # fill in external.data sublists : variable.id
+      external.data[[i]]$variable.id <-  varid[i]
+      
+      # fill in external.data sublists : input.id
+      external.data[[i]]$input.id <-  NA
+      
+      # fill in external.data sublists : align.method
+      external.data[[i]]$align.method <-  align_method
+      
+      # fill in external.data sublists : data
+      external.data[[i]]$data <-  obs[[i]]
+      
+      # fill in external.data sublists : obs
+      external.data[[i]]$obs <-  obs[[i]][[varn[i]]]
+      
+      # fill in external.data sublists : par
+      if(!is.null(par)){
```

Comment:
Hey @istfer I can't tell if this is a bug or a design feature. Right now this statement is set up so that if PAR is NULL, your functions will calculate the parameters for flux data. 

If the external.data isn't flux data, does it need a `par` value? The code is enforcing one, but I can't tell if it's necessary later on if priors are also provided.